File: Q1_Cpp/src/Shipment/Customer/CustomerCollection.cpp
#include "../include/Shipment/Customer/CustomerCollection.h"
#include <algorithm>

// Constructor for CustomerCollection
CustomerCollection::CustomerCollection( std::vector<std::shared_ptr<Customer>> customers)
:customers(customers) {
    // The vector 'customers' is automatically initialized
}

void CustomerCollection::addCustomer(const std::shared_ptr<Customer>& customer) {
    customers.push_back(customer);
}

std::shared_ptr<Customer> CustomerCollection::getCustomerById(const std::string& customerId) const {
    auto it = std::find_if(customers.begin(), customers.end(),
        [&customerId](const std::shared_ptr<Customer>& customer) {
            return customer->getId() == customerId;
        });

    if (it != customers.end()) {
        return *it;
    }
    return nullptr; // If customer is not found
}

std::vector<std::shared_ptr<Customer>> CustomerCollection::getAllCustomers() const {
    return customers; // Returning a copy of the customers vector
}

bool CustomerCollection::deleteCustomer(const std::string& customerId) {
    auto it = std::remove_if(customers.begin(), customers.end(),
        [&customerId](const std::shared_ptr<Customer>& customer) {
            return customer->getId() == customerId;
        });

    if (it != customers.end()) {
        customers.erase(it, customers.end());
        return true;
    }
    return false; // If no customer was removed
}

File: Q1_Java/src/main/java/Shipment/Customer/CustomerCollection.java
package Shipment.Customer; // Package for shipment and customer management

import java.util.ArrayList;
import java.util.List; // Importing list utilities


public class CustomerCollection {
    private List<Customer> customers; // List to store customers

    // Constructor for CustomerCollection
    public CustomerCollection() {
        customers = new ArrayList<>(); // Initializing the customer list
    }

    // Adds a customer to the collection
    public void addCustomer(Customer customer) {
        customers.add(customer); // Appending new customer
    }

    // Retrieves a customer by their ID
    public Customer getCustomerById(String customerId) {
        for (Customer customer : customers) {
            if (customer.getId().equals(customerId)) {
                return customer; // Found the customer
            }
        }
        return null; // No customer found
    }

    // Returns all customers in the collection
    public List<Customer> getAllCustomers() {
        return new ArrayList<>(customers); // Making a copy of customers list
    }

    
    // Deletes a customer by ID
    public boolean deleteCustomer(String customerId) {
        return customers.removeIf(customer -> customer.getId().equals(customerId)); // Removing customer by ID
        // removeIf method is part of List utilities and removes all elements in a list that satisfy the condition
    }

}
